### PR TITLE
Beef up the allow_in_graph docs

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127117

We make the following changes:
- most of the time when someone uses allow_in_graph, they actually
  wanted to make a custom op. We add a link to the custom ops landing
  page and explain the differences between allow_in_graph and custom
  ops.
- we warn people against using allow_in_graph footguns and document
  them.

Test Plan:
- tests

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: torch/_dynamo/decorators.py [hunk 1]
```diff
@@ -74,8 +74,40 @@ def assume_constant_result(fn):
...
+    The main use case for :func:`allow_in_graph()` is as an escape hatch for the compiler frontend:
+    if you know the function works w.r.t. to the downstream components of the compilation
+    stack (AOTAutograd and Inductor) but there is a Dynamo bug that prevents it from
+    symbolically introspecting the function properly, then one can decorate said function
+    with allow_in_graph to bypass Dynamo.
+
+    There are a number of restrictions on ``fn``: it must only accept inputs/outputs
+    of certain types (e.g. Tensor/int/bool/float/None and lists of the previous), and
+    all Tensors used inside of ``fn`` must be passed directly as inputs to ``fn``.
+    Failure to abide by these restrictions will result in undefined behavior.
```
Comment (jansel) [inline torch/_dynamo/decorators.py:94]: Let's be more precise about the restrictions.

File: torch/_dynamo/decorators.py [hunk 2]
```diff
@@ -74,8 +74,40 @@ def assume_constant_result(fn):
...
+    If you are trying to black-box a Python function for use with torch.compile
+    (aka something similar to torch.fx.wrap),
+    do not use this API. Instead, please create a custom operator:
+    https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU/edit
+
+    The main use case for :func:`allow_in_graph()` is as an escape hatch for the compiler frontend:
+    if you know the function works w.r.t. to the downstream components of the compilation
+    stack (AOTAutograd and Inductor) but there is a Dynamo bug that prevents it from
+    symbolically introspecting the function properly, then one can decorate said function
+    with allow_in_graph to bypass Dynamo.
```
Comment (jansel) [inline torch/_dynamo/decorators.py:89]: I don't think the usecase here is only to work around bugs.  I'd suggest adding something like:

| Mechanism      | Frontend (dynamo/symbolic_trace) | Backend (AOTAutograd) |
| -------------- | -------------------- | -------------------- |
| no decorator   | trace inside         | trace inside         | 
| allow_in_graph | opaque callable      | trace inside         |
| wrap (for FX) | opaque callable      | trace inside         |
| custom ops     | opaque callable      | opaque callable      |

Which makes the differences more clear.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/decorators.py (hunk 2 above)
Comment (zou3519) [reply torch/_dynamo/decorators.py:89]: Sounds good, I'll add a table

File: torch/_dynamo/decorators.py [hunk 3]
```diff
@@ -74,8 +74,40 @@ def assume_constant_result(fn):
...
     """
-    Customize which functions TorchDynamo will include in the generated
-    graph. Similar to `torch.fx.wrap()`.
+    Tells the compiler frontend (Dynamo) to skip symbolic introspection of the function
+    and instead directly write it to the graph when encountered.
+
+    If you are trying to black-box a Python function for use with torch.compile
+    (aka something similar to torch.fx.wrap),
+    do not use this API. Instead, please create a custom operator:
+    https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU/edit
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_dynamo/decorators.py:83]: short term: Landing page for custom ops (website) that links to existing tutorials, for more information see the gdoc.
longer term: (2.5) actually toss the gdoc into 1+ webpages

File: torch/_dynamo/decorators.py [hunk 4]
```diff
@@ -74,8 +74,40 @@ def assume_constant_result(fn):
 
 def allow_in_graph(fn):
     """
-    Customize which functions TorchDynamo will include in the generated
-    graph. Similar to `torch.fx.wrap()`.
+    Tells the compiler frontend (Dynamo) to skip symbolic introspection of the function
+    and instead directly write it to the graph when encountered.
+
+    If you are trying to black-box a Python function for use with torch.compile
+    (aka something similar to torch.fx.wrap),
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_dynamo/decorators.py:81]: Don't advertise custom ops as torch.fx.wrap, plop in the table, clarify that this is an fx.wrap if you're not using export

File: torch/_dynamo/decorators.py (hunk 2 above)
Comment (zou3519) [reply torch/_dynamo/decorators.py:89]: Jason: sometimes people want to execute stuff at compile time: allow_in_graph as a constant. e.g. we allow_in_graph einops (implementation is traceable: string parsing)

Jason is neutral on the rename.

**Comment by zou3519** [comment]
More followups:
# (make issue) zero-tensor meta
# mutate args can be optional (and be pessimistic)
# raw pybind: use warning instead of graph break (because the graph break message gets suppresssed)

**Comment by zou3519** [comment]
This should be ready for another pass through. I updated the PR with jansel's feedback and our offline discussion:
- removed mentions of torch.fx.wrap, because it's currently confusing (allow_in_graph is a torch.fx.wrap for Dynamo, but not for export today (because export today is Dynamo + AOTAutograd)).
- added more clarity on when to use allow_in_graph and when to use custom ops
- the docs page we link to is now a website

**Review by jansel — APPROVED** [review]

**Review by albanD — APPROVED** [review]
Very clear comparison with custom ops. Thanks!

**Pull request closed without merging**